src/Databases/DatabaseAtomic.cpp

```diff
@@ -112,6 +112,13 @@ StoragePtr DatabaseAtomic::detachTable(ContextPtr /* context */, const String &
     table_name_to_path.erase(name);
     detached_tables.emplace(table->getStorageID().uuid, table);
     not_in_use = cleanupDetachedTables();
+
+    if (!not_in_use.empty())
+    {
+        not_in_use.clear();
+        LOG_DEBUG(log, "Finished removing not used detached tables");
+    }
+
```

Comment:
I think this is a potential deadlock, see [here](https://s3.amazonaws.com/clickhouse-test-reports/64992/4a460760a7a55c3caeaf5537ac24437f90d9640d/integration_tests__tsan__[6_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel4_0.log).

Basically before your changes we first released the lock and then cleaned up the not used tables, while after your changes we clean up the tables while holding the lock. Maybe it is better to introduce a scope for the lock and do this outside of that scope.

Reply:
And maybe put some comment into the code to be obvious why it is necessary.

Reply:
Unfortently, Im not clear understand logic within dtors. However you right about diffrent order after my changes. I will fix it and add comments about the order. Thank you.

Reply:
Maybe let me try to explain it. It is a good example for the future too.

We have to mutexes:
1. `IDatabase::mutex` (let's call it `DM`), which is used to attach/detach tables etc.
2. `BackgroundSchedulePoolTaskInfo::exec_mutex` (let's call it `BM`), which is used when we do some operations on the task, e.g. activate or deactivate.

To execute a background task, we have to lock `BM`. In the background task of `StorageKafka`, we have to retrieve some table ids, so we call `IDatabase::tryGetTable`, which internally locks `DM`. Thus in this path `BM` is locked before `DM`. 

The dtor of `StorageKafka` destroys the `BackgroundSchedulePoolTaskHolder`s, which calls `BackgroundSchedulePoolTaskInfo::deactivate`, which locks `BM`. With you "old" changes, `~StorageKafka` is called while holding `DM`, thus in this way `DM` is locked before `BM`.

This means if we try to destroy a StorageKafka while one of its background tasks are trying to retrieve a table id, we would get into a deadlock.